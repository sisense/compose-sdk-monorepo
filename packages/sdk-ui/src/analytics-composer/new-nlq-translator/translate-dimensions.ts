import { Attribute, JaqlDataSourceForDto, JSONArray } from '@sisense/sdk-data';
import { NormalizedTable } from '../types.js';
import { createAttribute } from './common.js';

function isStringArray(value: JSONArray): value is string[] {
  return value.every((item) => typeof item === 'string');
}

export const translateDimensionsJSON = (
  dimensionsJSON: JSONArray,
  dataSource: JaqlDataSourceForDto,
  tables: NormalizedTable[],
): Attribute[] => {
  if (!dimensionsJSON) {
    return [];
  }
  if (!isStringArray(dimensionsJSON)) {
    throw new Error('Invalid dimensions JSON. Expected an array of strings.');
  }
  return dimensionsJSON.map((dimension) => {
    return createAttribute(dimension, dataSource, tables);
  });
};
