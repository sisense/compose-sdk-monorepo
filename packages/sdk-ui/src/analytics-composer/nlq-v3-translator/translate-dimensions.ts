import { Attribute } from '@sisense/sdk-data';

import { NlqTranslationError, NlqTranslationErrorContext, NlqTranslationResult } from '../types.js';
import { createAttributeFromName } from './common.js';
import { DimensionsInput, isStringArray } from './types.js';

/**
 * Translate an array of JSON objects to attributes
 *
 * @param input - DimensionsInput object`
 * @returns NlqTranslationResult<Attribute[]>
 */
export const translateDimensionsJSON = (
  input: DimensionsInput,
): NlqTranslationResult<Attribute[]> => {
  const { data: dimensionsJSON } = input;
  const { dataSource, tables } = input.context;

  if (!dimensionsJSON) {
    return { success: true, data: [] };
  }

  if (!isStringArray(dimensionsJSON)) {
    return {
      success: false,
      errors: [
        {
          category: 'dimensions',
          index: -1,
          input: dimensionsJSON,
          message: 'Invalid dimensions JSON. Expected an array of strings.',
        },
      ],
    };
  }

  const results: Attribute[] = [];
  const errors: NlqTranslationError[] = [];

  // Process each dimension and collect errors instead of throwing
  dimensionsJSON.forEach((dimensionString, index) => {
    const context: NlqTranslationErrorContext = {
      category: 'dimensions',
      index,
      input: dimensionString,
    };
    try {
      const attribute = createAttributeFromName(dimensionString, dataSource, tables);
      results.push(attribute);
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : 'Unknown error';
      errors.push({ ...context, message: errorMsg });
    }
  });

  return errors.length > 0 ? { success: false, errors } : { success: true, data: results };
};
