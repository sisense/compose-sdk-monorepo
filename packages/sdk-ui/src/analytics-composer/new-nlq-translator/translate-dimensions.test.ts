import { describe, it, expect } from 'vitest';
import { translateDimensionsJSON } from './translate-dimensions.js';

import {
  MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
  MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
} from '../__mocks__/mock-data-sources.js';

describe('translateDimensions', () => {
  it('should translate dimensions from JSON array of strings', () => {
    const mockDimensionsJSON = ['DM.Country.Country', 'DM.Brand.Brand', 'DM.Category.Category'];

    const dimensions = translateDimensionsJSON(
      mockDimensionsJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(dimensions).toHaveLength(3);
    expect(dimensions).toMatchSnapshot();
  });

  it('should translate dimensions with date levels', () => {
    const mockDimensionsJSON = ['DM.Commerce.Date.Years', 'DM.Commerce.Date.Months'];

    const dimensions = translateDimensionsJSON(
      mockDimensionsJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(dimensions).toHaveLength(2);
    expect(dimensions).toMatchSnapshot();
  });

  it('should return empty array when dimensionsJSON is null', () => {
    const dimensions = translateDimensionsJSON(
      null as any,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );
    expect(dimensions).toEqual([]);
  });

  it('should return empty array when dimensionsJSON is undefined', () => {
    const dimensions = translateDimensionsJSON(
      undefined as any,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );
    expect(dimensions).toEqual([]);
  });

  it('should return empty array when dimensionsJSON is false', () => {
    const dimensions = translateDimensionsJSON(
      false as any,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );
    expect(dimensions).toEqual([]);
  });

  it('should return empty array when dimensionsJSON is 0', () => {
    const dimensions = translateDimensionsJSON(
      0 as any,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );
    expect(dimensions).toEqual([]);
  });

  it('should return empty array when dimensionsJSON is empty string', () => {
    const dimensions = translateDimensionsJSON(
      '' as any,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );
    expect(dimensions).toEqual([]);
  });

  it('should translate empty array to empty array', () => {
    const dimensions = translateDimensionsJSON(
      [],
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );
    expect(dimensions).toEqual([]);
  });

  it('should throw error for array containing non-string values', () => {
    const mockDimensionsJSON = ['DM.Country.Country', 123, 'DM.Brand.Brand'] as any;

    expect(() => {
      translateDimensionsJSON(
        mockDimensionsJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
    }).toThrow('Invalid dimensions JSON. Expected an array of strings.');
  });

  it('should throw error for array containing object values', () => {
    const mockDimensionsJSON = [
      'DM.Country.Country',
      { invalid: 'object' },
      'DM.Brand.Brand',
    ] as any;

    expect(() => {
      translateDimensionsJSON(
        mockDimensionsJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
    }).toThrow('Invalid dimensions JSON. Expected an array of strings.');
  });

  it('should throw error for array containing null values', () => {
    const mockDimensionsJSON = ['DM.Country.Country', null, 'DM.Brand.Brand'] as any;

    expect(() => {
      translateDimensionsJSON(
        mockDimensionsJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
    }).toThrow('Invalid dimensions JSON. Expected an array of strings.');
  });

  it('should throw error for array containing boolean values', () => {
    const mockDimensionsJSON = ['DM.Country.Country', true, 'DM.Brand.Brand'] as any;

    expect(() => {
      translateDimensionsJSON(
        mockDimensionsJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
    }).toThrow('Invalid dimensions JSON. Expected an array of strings.');
  });

  it('should throw error for array containing array values', () => {
    const mockDimensionsJSON = ['DM.Country.Country', ['nested', 'array'], 'DM.Brand.Brand'] as any;

    expect(() => {
      translateDimensionsJSON(
        mockDimensionsJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
    }).toThrow('Invalid dimensions JSON. Expected an array of strings.');
  });
});
