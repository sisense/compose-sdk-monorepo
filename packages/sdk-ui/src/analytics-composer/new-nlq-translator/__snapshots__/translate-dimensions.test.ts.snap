// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`translateDimensions > should translate dimensions from JSON array of strings 1`] = `
[
  {
    "__serializable": "DimensionalAttribute",
    "composeCode": "DM.Country.Country",
    "dataSource": {
      "address": "LocalHost",
      "id": "localhost_aSampleIAAaECommerce",
      "live": false,
      "title": "Sample ECommerce",
    },
    "description": "",
    "expression": "[Country.Country]",
    "name": "Country",
    "type": "text-attribute",
  },
  {
    "__serializable": "DimensionalAttribute",
    "composeCode": "DM.Brand.Brand",
    "dataSource": {
      "address": "LocalHost",
      "id": "localhost_aSampleIAAaECommerce",
      "live": false,
      "title": "Sample ECommerce",
    },
    "description": "",
    "expression": "[Brand.Brand]",
    "name": "Brand",
    "type": "text-attribute",
  },
  {
    "__serializable": "DimensionalAttribute",
    "composeCode": "DM.Category.Category",
    "dataSource": {
      "address": "LocalHost",
      "id": "localhost_aSampleIAAaECommerce",
      "live": false,
      "title": "Sample ECommerce",
    },
    "description": "",
    "expression": "[Category.Category]",
    "name": "Category",
    "type": "text-attribute",
  },
]
`;

exports[`translateDimensions > should translate dimensions with date levels 1`] = `
[
  {
    "__serializable": "DimensionalLevelAttribute",
    "composeCode": "DM.Commerce.Date.Years",
    "dataSource": {
      "address": "LocalHost",
      "id": "localhost_aSampleIAAaECommerce",
      "live": false,
      "title": "Sample ECommerce",
    },
    "description": "",
    "expression": "[Commerce.Date]",
    "format": "yyyy",
    "granularity": "Years",
    "name": "Date",
    "type": "datelevel",
  },
  {
    "__serializable": "DimensionalLevelAttribute",
    "composeCode": "DM.Commerce.Date.Months",
    "dataSource": {
      "address": "LocalHost",
      "id": "localhost_aSampleIAAaECommerce",
      "live": false,
      "title": "Sample ECommerce",
    },
    "description": "",
    "expression": "[Commerce.Date]",
    "format": "yyyy-MM",
    "granularity": "Months",
    "name": "Date",
    "type": "datelevel",
  },
]
`;
